added support of insert from url option in cover block

akasunil opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Currently it isn't possible to insert images or video via url for cover block. 

## What?
Fixes [10853](https://github.com/WordPress/gutenberg/issues/10853)

## Testing Instructions
Please include step by step instructions on how to test this PR.
1. Go to post edit.
2. Insert a Cover Block.

## Screenshots or screencast <!-- if applicable -->
<img width="754" alt="CleanShot 2022-06-23 at 17 22 50@2x" src="https://user-images.githubusercontent.com/61308756/175292681-1c8f5424-ede4-45e1-938e-1c52c1d3469d.png">
<img width="570" alt="CleanShot 2022-06-23 at 17 24 06@2x" src="https://user-images.githubusercontent.com/61308756/175292660-5d5b7f97-e149-46cd-8e81-e16bc5f31972.png">

akasunil committed:
added support of insert from url option in cover block

akasunil commented:
@Mamaduka can you please review this PR.

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

Mamaduka commented:
Thanks for working on this, @sunil25393.

Unfortunately, using regex to detect media type won't work with URLs that don't have extensions attached. E.g. - `https://avatars.githubusercontent.com/u/61308756?s=200&v=4`.

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
added function to get content type of url

akasunil commented:
> Thanks for working on this, @sunil25393.
> 
> Unfortunately, using regex to detect media type won't work with URLs that don't have extensions attached. E.g. - `https://avatars.githubusercontent.com/u/61308756?s=200&v=4`.

Have addressed feedback. please review it again.

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil commented:
Any more feedbacks ?

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil commented:
@Mamaduka can we merge this PR if there isnt any more feedback. 

Thank you.

Mamaduka commented:
Hi, @sunil25393

Sorry, I didn't get a chance to test or review changes here.

Quick notes:

* We shouldn't add a new dependency to use `fetch`. The `window.fetch` should work here. Example in core - https://github.com/WordPress/gutenberg/blob/trunk/packages/api-fetch/src/index.js#L101.
* Have you tested that `getContentTypeFromUrl` doesn't produce CORS errors in the console?

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Removed fetch dependency

akasunil commented:
Hi @Mamaduka 

I have addressed feedback, and tested to make sure that isn't any CORS errors in the console.

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Removed dependancy from package.json

akasunil commented:
Waiting for PR review.

Mamaduka commented:
I pinged some folks who recently worked on the Cover block for the reviews.

draganescu commented:
Hi @sunil25393,

Generally this works well if the host that we're linking to allows it. For instance I can easily run into:

```
Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://jsoncompare.org/LearningContainer/SampleFiles/Video/MP4/Sample-MP4-Video-File-Download.mp4. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing). Status code: 200.
```

But this other host `https://vod-progressive.akamaized.net/exp=1661274774~acl=%2Fvimeo-prod-skyfire-std-us%2F01%2F3603%2F14%2F368016235%2F1522131611.mp4~hmac=1558863b3bea9ef9660af2cfd3e6e8adce91ae3310367141d718837f2f1b198f/vimeo-prod-skyfire-std-us/01/3603/14/368016235/1522131611.mp4` works just fine.

This inconsistency has to be treated better, not sure how, but at this point as a user I have no idea why it does not work.

I have tested with URLs that allow remote linking and the features work as expected. We just need to figure out what to do if:

- the media is being blocked by CORS
- the media is not image or video, and we supply a link to a pdf - in the other cases we limit the "picker" to media files, but a user may paste any URL to any kind of file.

akasunil commented:
i see, do you know how we do for external image on image block ?

draganescu commented:
We don't do it there perfectly either. We can link to [a pdf file](https://www.w3.org/WAI/ER/tests/xhtml/testfiles/resources/pdf/dummy.pdf) and get an empty block. 

Nevertheless in the image block we don't have to deal with CORS because the `<img>` tag used in the block's rendering is [not subject](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#what_requests_use_cors) to CORS policies, hence we don't do anything.

In the case of the cover block it is in your implementation of `getContentTypeFromUrl`  where the CORS problem is triggered as we cannot read cross origin files by default. The CDN example allows it by default because its their policy but the other domain has the standard policy in place. Since we cannot detect the type because CORS prevents us, we would need a way to ask the user what they're linking to.

Added a needs design feedback for this.

akasunil commented:
Thank you @draganescu. as you said, we should add field for content type of link.

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

akasunil committed:
Merge branch 'WordPress:trunk' into cover-block-insert-from-url-support

richtabor commented:
> Added a needs design feedback for this.

Do we just need a better alert/fallback state when there's a failure?

draganescu commented:
@richtabor no, we also need to ask what kind of file are we linking to.
